### PR TITLE
Generating correctly waypoints ahead when car is not moving.

### PR DESCRIPTION
**Pull request by vguerra**
Currently the car publishes a lane ahead starting at waypoint 199
and eventhough the car moves the lane does not shift correctly.

Unite test will come in next commit.

### CODE REVIEW

**Review by asimonov — APPROVED** [review]
lgtm

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 1]
```diff
@@ -113,14 +115,15 @@ def _closest_waypoint_index(self):
                     self.current_pose.position.y)
         
         if self.current_waypoint_ahead is None:
-            possible_waypoint_indices = self._get_waypoint_indices(0, LOOKAHEAD_WPS)
+            possible_waypoint_indices = self._get_waypoint_indices(0,
+                                                                   self.len_base_waypoints)
```
Comment (avrabe) [inline ros/src/waypoint_updater/waypoint_updater.py:119]: There might be a possible race condition if the waypoints_cb() function is called while the waypoints are updated and published.

Comment (vguerra) [reply ros/src/waypoint_updater/waypoint_updater.py:119]: yes.. that crossed my mind as well. I will address that in a future PR 👍

**Comment by asimonov** [comment]
Is it not one python file without threaded code? I would have thought there is no parallelism within it. Unless some of the rospy calls setup threads. Good question though!

_____________________________
Kind Regards,
Alexey Simonov

**Comment by avrabe** [comment]
The rospy.init_node creates the communication threads (https://github.com/ros/ros_comm/blob/9e31e644ac88e60b75fca5018181175942195009/clients/rospy/src/rospy/impl/masterslave.py#L222)
Init_node -> start_node -> RosHandler -> threadin.Thread

**Comment by asimonov** [comment]
thank you Ralf.

that got me thinking I wanna read up more on concurrency in ROS.
looks like this is how roscpp deals with it: 
http://wiki.ros.org/roscpp/Overview/Callbacks%20and%20Spinning

But rospy best practice of handling callbacks is hard to find. 
Here is some evidence that suggests locking is needed:
https://answers.ros.org/question/115251/a-rospy-subscriber-callback-thread-locking-question/
https://answers.ros.org/question/9543/rospy-threading-model/

**Comment by avrabe** [comment]
You're welcome. Yes, threading.Lock is a good way. I used it for the dbw unit test for the similar reasons.On 10 Oct 2017 1:39 PM, Alexey <notifications@github.com> wrote:thank you Ralf.
that got me thinking I wanna read up more on concurrency in ROS.
looks like this is how roscpp deals with it:
http://wiki.ros.org/roscpp/Overview/Callbacks%20and%20Spinning
But rospy best practice of handling callbacks is hard to find.
Here is some evidence that suggests locking is needed:
https://answers.ros.org/question/115251/a-rospy-subscriber-callback-thread-locking-question/
https://answers.ros.org/question/9543/rospy-threading-model/

**Pull request merged**